### PR TITLE
Adds R4 Core Project

### PR DESCRIPTION
**Pull request by brendankowitz**
## Description
* Adds Shared projects for:
  * Fhir.[Stu3/R4].Core
  * Fhir.[Stu3/R4].Core.Tests
  * Fhir.Tests.Common
* Adds R4 Core project
* Adds R4 Core Tests project
* Removes Stu3 from Fhir.Tests.Common project
* Removes Stu3 from Fhir.Core.UnitTests project

## Related issues
Addresses [AB#69183](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/69183).

## Testing
Refactors Core unit tests

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Definition/SearchParameterDefinitionBuilder.cs [hunk 1]
```diff
@@ -100,8 +100,11 @@ private static bool ShouldExcludeEntry(string resourceType, string searchParamet
         {
             return (resourceType == ResourceType.DomainResource.ToString() && searchParameterName == "_text") ||
                 (resourceType == ResourceType.Resource.ToString() && searchParameterName == "_content") ||
-                (resourceType == ResourceType.Resource.ToString() && searchParameterName == "_query") ||
-                (resourceType == ResourceType.DataElement.ToString() && (searchParameterName == "objectClass" || searchParameterName == "objectClassProperty"));
+                (resourceType == ResourceType.Resource.ToString() && searchParameterName == "_query")
+#if !R4
```
Comment (brandonpollett) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Definition/SearchParameterDefinitionBuilder.cs]: Lets make this `#if Stu3`

File: src/Microsoft.Health.Fhir.R4.Core/Features/Conformance/ListedCapabilityStatement.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+using System.Collections.Generic;
+using EnsureThat;
+using Hl7.Fhir.Model;
+using Microsoft.Health.Fhir.Core.Extensions;
+using static Hl7.Fhir.Model.CapabilityStatement;
+
+namespace Microsoft.Health.Fhir.Core.Features.Conformance
+{
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Usage", "CA2227:Collection properties should be read only", Justification = "This is a DTO-style class")]
+    public sealed class ListedCapabilityStatement : IListedCapabilityStatement
```
Comment (brandonpollett) [inline src/Microsoft.Health.Fhir.R4.Core/Features/Conformance/ListedCapabilityStatement.cs]: Could we have a base version of this and specific implementations?

Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.R4.Core/Features/Conformance/ListedCapabilityStatement.cs]: I'll do this in a separate PR

File: src/Microsoft.Health.Fhir.Core/Models/ModelInfoProvider.cs [hunk 3]
```diff
@@ -14,6 +14,10 @@ public static class ModelInfoProvider
     {
         private static IModelInfoProvider _modelInfoProvider;
 
+        public static FhirSpecification Version => _modelInfoProvider?.Version ?? FhirSpecification.Stu3;
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Core/Models/ModelInfoProvider.cs]: nit: maybe throw exception rather than automatically default to STU3?

Comment (jackliums) [reply src/Microsoft.Health.Fhir.Core/Models/ModelInfoProvider.cs]: (perhaps call EnsureProvider() before returning the instance here and below.)

**Comment by jackliums** [comment]
Is the AllCapabilities file correct? I see some extensions and what not which doesn't seem like it should be there:

```
            "extension": [
                {
                    "url": "http://hl7.org/fhir/StructureDefinition/capabilitystatement-websocket",
                    "valueString": "http://test.fhir.org/r4/websockets"
                },
                {
                    "extension": [
                        {
                            "url": "name",
                            "valueString": "Fetch Patient Alerts"
                        },
                        {
                            "url": "activity",
                            "valueString": "patient-view"
                        },
                        {
                            "url": "preFetchOptional",
                            "valueString": "Patient/{{Patient.id}}"
                        }
                    ],
                    "url": "http://fhir-registry.smarthealthit.org/StructureDefinition/cds-activity"
                },
                {
                    "extension": [
                        {
                            "url": "name",
                            "valueString": "Get Terminology Information"
                        },
                        {
                            "url": "activity",
                            "valueString": "code-view"
                        }
                    ],
                    "url": "http://fhir-registry.smarthealthit.org/StructureDefinition/cds-activity"
                },
                {
                    "extension": [
                        {
                            "url": "name",
                            "valueString": "Get identifier Information"
                        },
                        {
                            "url": "activity",
                            "valueString": "identifier-view"
                        }
                    ],
                    "url": "http://fhir-registry.smarthealthit.org/StructureDefinition/cds-activity"
                }
            ],
```

**Comment by brendankowitz** [comment]
> Is the AllCapabilities file correct? I see some extensions and what not which doesn't seem like it should be there:
> 
> ```
>             "extension": [
>                 {
>                     "url": "http://hl7.org/fhir/StructureDefinition/capabilitystatement-websocket",
>                     "valueString": "http://test.fhir.org/r4/websockets"
>                 },
>                 {
>                     "extension": [
>                         {
>                             "url": "name",
>                             "valueString": "Fetch Patient Alerts"
>                         },
>                         {
>                             "url": "activity",
>                             "valueString": "patient-view"
>                         },
>                         {
>                             "url": "preFetchOptional",
>                             "valueString": "Patient/{{Patient.id}}"
>                         }
>                     ],
>                     "url": "http://fhir-registry.smarthealthit.org/StructureDefinition/cds-activity"
>                 },
>                 {
>                     "extension": [
>                         {
>                             "url": "name",
>                             "valueString": "Get Terminology Information"
>                         },
>                         {
>                             "url": "activity",
>                             "valueString": "code-view"
>                         }
>                     ],
>                     "url": "http://fhir-registry.smarthealthit.org/StructureDefinition/cds-activity"
>                 },
>                 {
>                     "extension": [
>                         {
>                             "url": "name",
>                             "valueString": "Get identifier Information"
>                         },
>                         {
>                             "url": "activity",
>                             "valueString": "identifier-view"
>                         }
>                     ],
>                     "url": "http://fhir-registry.smarthealthit.org/StructureDefinition/cds-activity"
>                 }
>             ],
> ```

Yep, removed these

File: src/Microsoft.Health.Fhir.Shared.Tests/Deserializers.cs [hunk 4]
```diff
@@ -7,15 +7,14 @@
...
 using Microsoft.Health.Fhir.Core.Extensions;
-using Microsoft.Health.Fhir.Core.Features;
 using Microsoft.Health.Fhir.Core.Features.Persistence;
 using Microsoft.Health.Fhir.Core.Models;
 
 namespace Microsoft.Health.Fhir.Tests.Common
 {
     public static class Deserializers
     {
-        private static readonly FhirJsonParser JsonParser = new FhirJsonParser(DefaultParserSettings.Settings);
```
[the remark this answers is not in the document]
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Shared.Tests/Deserializers.cs]: It is, but I'm going to keep our settings object

File: src/Microsoft.Health.Fhir.Shared.Core/AssemblyInfo.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("DynamicProxyGenAssembly2")]
```
[the remark this answers is not in the document]
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Shared.Core/AssemblyInfo.cs]: Dup, removing

**Pull request merged**